### DISCUSSION TITLE
How do I get the current route name (i.e. with placeholders like [slug] intact)?

### DISCUSSION THREAD

**Discussion by billliu1992** [vercel/next.js#56973] [Help]
### Summary

Hi Next.JS community,

I'm trying to get the current route name, NOT the path name. If you look at the [useParams docs](https://nextjs.org/docs/app/api-reference/functions/use-params), you'll see there is a "Route" like "app/shop/[slug]/page.js" and there is a path like "/shop/1". I want the former ("app/shop/[slug]/page.js") NOT the latter. I know `usePathname` gives me the latter, but I've been unable to find a way to get the former.

I know it's possible in the old Page Router, as `pathname` of the [router object](https://nextjs.org/docs/pages/api-reference/functions/use-router#router-object) gives you the filepath. The docs say `usePathname` in the App Router is a replacement for `pathname` in the Page Router, but it does something completely different: when I console.log the return value of `usePathname` inside a dynamic segment `/test/[slug]/page.tsx`, I get something like `/test/hi` instead of `/test/[slug]/page.tsx`.

I want to do so because I want to specifically monitor errors and query rate per page. Ideally, I want to be able to group error rates and query rates among the route name so I can easily know if a specific page is throwing a lot of errors. If I only have the pathname, I can't do so. An added bonus would be being able to detect intercepted routes (like, we can figure out if we're in the route `/test/(..)/hi/[slug]` vs `/hi/[slug]`).

I've scoured the docs, and made multiple searches across the internet, to no avail.

Not sure if it helps, but I'm currently on Next.js version "13.5.4"

Thanks!

**Comment by FanManutd** [comment]
#60882

**Comment by julianklumpers** [comment]
One way to achieve this could be by reversing the `params` object and interpolate a new `pathName` with it.

```ts
  const pathname = usePathname();
  const params = useParams();

  const reverseParamsObject = Object.entries(params).reduce(
    (prev, [key, val]) => ({ ...prev, [val]: key }),
    {} as typeof params,
  );
  const fileRoutePath = pathname
    .split('/')
    .reduce(
      (prev, cur, idx, arr) =>
        `${prev}${reverseParamsObject[cur] ? `[${reverseParamsObject[cur]}]` : cur}${arr.length !== idx + 1 ? '/' : ''}`,
      '',
    );

/** file structure
 * app
 *  posts
 *    page.tsx
 *    [domain]
 *      page.tsx
 *      [id]
 *        page.tsx
 *  layout.tsx
 */

// input: /posts/tech/123
// output: /posts/[domain]/[id]
```

**Reply by billliu1992** [reply]
This doesn't work in cases where multiple parameters have the same value, or if the parameters have the same value as another part of the URL. For example:

Original route: `/test/[param1]/[param2]`
Parameters: `{ 'param1': 'test', 'param2': 'test'}`

We'd expect the original (`/test/[param1]/[param2]`), but with your code we'd actually get something like `/[param1]/[param1]/[param1]`

Your solution also doesn't account for catch-all segments.

I'm doing something similar as a workaround, but there's no perfect solution. My use-case is for monitoring, and not even the provided OpenTelemetry traces have this info for subsequent navigations.

**Comment by julianklumpers** [comment]
@billliu1992 you are right, I adjusted my code. I hope you get something useful out of it. In my case, I needed this for generating breadcrumb links with custom names.

I'm doing it backwards and reverse the path at the end, because URL parameters usually come at the end of a URL, but that is just an assumption. i agree it is not the perfect solution, but for me, it works :)
Note: it is not the prettiest code, I made it quickly for test purposes.

```ts
const getFileRoutePath = (pathname, params) => {
  let paramCount = 0;
  const reverseParamsObject = Object.entries(params).reduceRight(
    (prev, [key, val]) => {
      if (Array.isArray(val)) {
        return {
          ...prev,
          ...val.reduceRight((_prev, cur, _idx) => {
            _prev[`${cur}-${paramCount}`] = `[...${key}]`;
            paramCount += 1;
            return _prev;
          }, {}),
        };
      }

      prev[`${val}-${paramCount}`] = `[${key}]`;
      paramCount += 1;
      return prev;
    },
    {} as typeof params
  );

  let interpolateCount = 0;
  const fileRoutePath = pathname.split('/').reduceRight((prev, cur) => {
    const lookup = reverseParamsObject[`${cur}-${interpolateCount}`];

    if (prev.endsWith(`${lookup}${'/'}`)) {
      interpolateCount += 1
      return prev;
    }

    if (lookup) {
      prev += lookup;
      interpolateCount += 1;
    } else {
      prev += cur;
    }

    return prev + '/';
  }, '').slice(0, -1).split('/').reverse().join('/');

  return fileRoutePath;
};

  const pathname = '/posts/test/test/123/foo/bar';
  const params = {
    param1: 'test',
    param2: 'test',
    id: '123',
    slug: ['foo', 'bar'],
  };

getFileRoutePath(pathname, params)
// output: '/posts/[param1]/[param2]/[id]/[...slug]'

const pathname = '/test/test/test/test/test/123/one/two/three';
const params = {
  param1: 'test',
  param2: 'test',
  id: 123,
  slug: ['one', 'two', 'three'],
};

getFileRoutePath(pathname, params);
// output: '/test/test/test/[param1]/[param2]/[id]/[...slug]'
```
